subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java

```diff
@@ -72,25 +74,30 @@
     private final StringInterner stringInterner;
     private final HashCode zipHasherConfigurationHash;
 
-    private ClasspathFingerprintingStrategy(String identifier, NonJarFingerprintingStrategy nonZipFingerprintingStrategy, ResourceHasher classpathResourceHasher, ResourceFilter classpathResourceFilter, ResourceSnapshotterCacheService cacheService, StringInterner stringInterner) {
+    private ClasspathFingerprintingStrategy(String identifier, NonJarFingerprintingStrategy nonZipFingerprintingStrategy, ResourceHasher classpathResourceHasher, ClasspathResourceFilters classpathResourceFilters, ResourceSnapshotterCacheService cacheService, StringInterner stringInterner) {
         super(identifier);
         this.nonZipFingerprintingStrategy = nonZipFingerprintingStrategy;
-        this.classpathResourceFilter = classpathResourceFilter;
+        this.classpathResourceFilter = classpathResourceFilters.getResourceFilter();
         this.classpathResourceHasher = classpathResourceHasher;
         this.cacheService = cacheService;
         this.stringInterner = stringInterner;
-        this.zipHasher = new ZipHasher(classpathResourceHasher, classpathResourceFilter);
+        if (classpathResourceFilters instanceof RuntimeClasspathResourceFilters) {
+            RuntimeClasspathResourceFilters runtimeClasspathResourceFilters = (RuntimeClasspathResourceFilters) classpathResourceFilters;
+            this.zipHasher = new ZipHasher(classpathResourceHasher, classpathResourceFilter, runtimeClasspathResourceFilters.getManifestAttributeResourceFilter(), runtimeClasspathResourceFilters.getManifestPropertyResourceFilter());
+        } else {
+            this.zipHasher = new ZipHasher(classpathResourceHasher, classpathResourceFilter);
+        }
```

Comment:
I wasn't concerned about this being upside down because it's a private constructor and the callees are the static constructor methods in the same class. Though I should at least be asserting that `classpathResourceFilters` is `RuntimeClasspathResourceFilters` in that static method I guess?

I went with this approach because there are other classes such as `DefaultClasspathFingerprinter` where overloading constructors, blowing out fields, etc. doesn't make sense, the strategy/hasher is the only place we need that along that part of the object graph. 

I'll tinker with it and see if I can come up with something a little cleaner.

Reply:
Got rid of the overloaded constructor to make the `FILTER_NOTHING` more obvious.

Reply:
I think `ClasspathResourceFilters` is still useful, given the other classes in that object graph that needs to pass these around and avoids needing currently three, and in future possibly more filter types.

Reply:
Let's move the `if` out of here, to the two methods calling this constructor. We have exactly two places this gets called from the same class, so no need for this complexity. If we do that, `ClasspathResourceFilters` becomes unnecessary.

Reply:
The `ClasspathResourceFilters` wrapper is to avoid constructor argument blow out where this is passed through as a depenency - `VirtualFileSystemServices` and `DefaultClasspathFingerprinter` specifically. Would you prefer we add two additional constructor parameters (and later more should normalization be further enhanced)?

Reply:
Or are you suggesting `RuntimeClasspathResourceFilters` is unnecessary and `ClasspathResourceFilters` should carry it with the `FILTER_NOTHING` filters for classpath normalization?